PR: https://github.com/odoo/odoo/pull/

From: db92ca5fce6477dc9f65e904193651403b98b940
From: jem-odoo
Date: 2019-05-24 09:20:06

Structural Changes: 4
Total Changes: 96

[IMP] project,hr_timesheet: move analytic account business to project

When we remove the inheritS between project and analytic account, we move the
analytic_accound_id field in timesheet because it was the only use case for
a project to be linked to an analytic account.
This commit prepares new cost origin for a project (other that timesheet).
Register costs on project will required an analytic account to keep analytic
costs tracking. The analytic account management is now moved to project, like
it was in 10.0, but without the inheritS.

Task-1911581

================================= pseudo patch: =================================

--- a/addons/project/models/__init__.py
+++ b/addons/project/models/__init__.py
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
+from . import analytic_account
 from . import project
 from . import res_config_settings
 from . import res_partner

--- a/None
+++ b/addons/project/models/analytic_account.py
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+
+class AccountAnalyticAccount(models.Model):
+    _inherit = 'account.analytic.account'
+    _description = 'Analytic Account'
+
+    project_ids = fields.One2many('project.project', 'analytic_account_id', string='Projects')
+    project_count = fields.Integer("Project Count", compute='_compute_project_count')
+
+    @api.multi
+    @api.depends('project_ids')
+    def _compute_project_count(self):
+        project_data = self.env['project.project'].read_group([('analytic_account_id', 'in', self.ids)], ['analytic_account_id'], ['analytic_account_id'])
+        mapping = {m['analytic_account_id'][0]: m['analytic_account_id_count'] for m in project_data}
+        for account in self:
+            account.project_count = mapping.get(account.id, 0)
+
+    @api.multi
+    def unlink(self):
+        projects = self.env['project.project'].search([('analytic_account_id', 'in', self.ids)])
+        has_tasks = self.env['project.task'].search_count([('project_id', 'in', projects.ids)])
+        if has_tasks:
+            raise UserError(_('Please remove existing tasks in the project linked to the accounts you want to delete.'))
+        return super(AccountAnalyticAccount, self).unlink()
+
+    @api.multi
+    def action_view_projects(self):
+        kanban_view_id = self.env.ref('project.view_project_kanban').id
+        result = {
+            "type": "ir.actions.act_window",
+            "res_model": "project.project",
+            "views": [[kanban_view_id, "kanban"], [False, "form"]],
+            "domain": [['analytic_account_id', '=', self.id]],
+            "context": {"create": False},
+            "name": "Projects",
+        }
+        if len(self.project_ids) == 1:
+            result['views'] = [(False, "form")]
+            result['res_id'] = self.project_ids.id
+        return result

--- a/addons/project/models/project.py
+++ b/addons/project/models/project.py
@@ -80,13 +80,6 @@ class Project(models.Model):
         values['alias_defaults'] = {'project_id': self.id}
         return values
 
-    @api.multi
-    def unlink(self):
-        for project in self:
-            if project.tasks:
-                raise UserError(_('You cannot delete a project containing tasks. You can either archive it or first delete all of its tasks.'))
-        return super(Project, self).unlink()
-
     def _compute_attached_docs_count(self):
         Attachment = self.env['ir.attachment']
         for project in self:
@@ -180,6 +173,9 @@ class Project(models.Model):
     partner_id = fields.Many2one('res.partner', string='Customer', auto_join=True, tracking=True)
     company_id = fields.Many2one('res.company', string='Company', required=True, default=lambda self: self.env.company_id)
     currency_id = fields.Many2one('res.currency', related="company_id.currency_id", string="Currency", readonly=True)
+    analytic_account_id = fields.Many2one('account.analytic.account', string="Analytic Account", copy=False, ondelete='set null',
+        help="Analytic account to which this project is linked for financial management."
+             "Use an analytic account to record cost and revenue on your project.")
 
     favorite_user_ids = fields.Many2many(
         'res.users', 'project_favorite_user_rel', 'project_id', 'user_id',
@@ -324,6 +320,21 @@ class Project(models.Model):
                 project.message_subscribe(project.partner_id.ids)
         return res
 
+    @api.multi
+    def unlink(self):
+        # Check project is empty
+        for project in self:
+            if project.tasks:
+                raise UserError(_('You cannot delete a project containing tasks. You can either archive it or first delete all of its tasks.'))
+        # Delete the empty related analytic account
+        analytic_accounts_to_delete = self.env['account.analytic.account']
+        for project in self:
+            if project.analytic_account_id and not project.analytic_account_id.line_ids:
+                analytic_accounts_to_delete |= project.analytic_account_id
+        result = super(Project, self).unlink()
+        analytic_accounts_to_delete.unlink()
+        return result
+
     @api.multi
     def message_subscribe(self, partner_ids=None, channel_ids=None, subtype_ids=None):
         """ Subscribe to all existing active tasks when subscribing to a project """
@@ -383,6 +394,31 @@ class Project(models.Model):
         action_context.pop('group_by', None)
         return dict(action, context=action_context)
 
+    # ---------------------------------------------------
+    #  Business Methods
+    # ---------------------------------------------------
+
+    @api.model
+    def _create_analytic_account_from_values(self, values):
+        analytic_account = self.env['account.analytic.account'].create({
+            'name': values.get('name', _('Unknown Analytic Account')),
+            'company_id': values.get('company_id', self.env.user.company_id.id),
+            'partner_id': values.get('partner_id'),
+            'active': True,
+        })
+        return analytic_account
+
+    @api.multi
+    def _create_analytic_account(self):
+        for project in self:
+            analytic_account = self.env['account.analytic.account'].create({
+                'name': project.name,
+                'company_id': project.company_id.id,
+                'partner_id': project.partner_id.id,
+                'active': True,
+            })
+            project.write({'analytic_account_id': analytic_account.id})
+
     # ---------------------------------------------------
     # Rating business
     # ---------------------------------------------------
